### PR TITLE
559091 -  generic API for service registry

### PR DESCRIPTION
**Pull request by eparovyshnaya**
first design version and raw implementation

### CODE REVIEW

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Configuration.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * Contributors:
+ *     ArSysOp - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.passage.lic.api.internal.registry;
+
+/**
+ * 
+ * @since 0.7
+ */
+public interface Configuration {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Configuration.java]: Any descriptive comments?

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Registry.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * @param <S> sub type of {@linkplain Service}
+ * @since 0.7
+ */
+public interface Registry<S extends Service<?>> {
+
+	boolean hasService(ServiceId id);
+
+	S service(ServiceId id);
+
+	Collection<S> services();
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Registry.java]: List? I expect it ordered but cannot explain why.

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Configuration.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+	public static final class Empty implements Configuration {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Configuration.java]: Is this needed?

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Service.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ *     ArSysOp - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.passage.lic.api.internal.registry;
+
+/**
+ * 
+ * @param <C> sub type of {@linkplain Configuration}
+ * @since 0.7
+ */
+public interface Service<C extends Configuration> {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Service.java]: May be 
```suggestion
public interface Service<I extends Identifier, C extends Configuration> {
```

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/ServiceId.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ *******************************************************************************/
+package org.eclipse.passage.lic.api.internal.registry;
+
+/**
+ * Implementation is expected to represent <i>data class</i>: with
+ * {@code hashCode} and {@code equals} overloaded basing on enclosed data.
+ * 
+ * @since 0.7
+ */
+public interface ServiceId {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/ServiceId.java]: `Identifier` ?

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Registry.java (hunk 2 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Registry.java]: In general case order will be unpredictable eitherway. 

Indeed, default implementation of, say, `RuntimeRegistry` allows to install a particular implementation of physical storage (say, `TreeMap`), which potentially can be implicitely used for service ordering. 

But on the current depth of api design detalisation _ordering_  is not the feature of registries. It should 
 - either become one and beget more complex interfaces - but benefits are not clear
 - or copletely kept for clients - and in this case it looks better to precisely declare "do not rely on any order".

Does this convince you?

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Configuration.java (hunk 3 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Configuration.java]: Most of service are not configurable. I expect this implementation will be the most usable one.

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/ServiceId.java (hunk 5 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/ServiceId.java]: FQN will be `...registry.Identifier` which is not the case, it should be identifier for _service_. 

`ServiceIdentifier`?

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Configuration.java (hunk 1 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Configuration.java]: Done

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Service.java (hunk 4 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Service.java]: Done

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Registry.java (hunk 2 above)
Comment (ruspl-afed) [reply bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Registry.java]: ok

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/ServiceId.java (hunk 5 above)
Comment (ruspl-afed) [reply bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/ServiceId.java]: `ServiceIdentifier` looks better fr me

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Service.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * </p>
+ * 
+ * @param <I> sub type of {@linkplain ServiceId}
+ * @param <C> sub type of {@linkplain Configuration}
+ */
+public interface Service<I extends ServiceId, C extends Configuration> {
+
+	I id();
+
+	C configuration();
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/api/internal/registry/Service.java]: Let's remove this, the configuration should not belong to service

**Comment by eparovyshnaya** [comment]
Closes [559091](https://bugs.eclipse.org/bugs/show_bug.cgi?id=559091)

**Pull request merged**